### PR TITLE
Correct Null ecosystem Display

### PR DESCRIPTION
**Pull request by fundthmcalculus**
If ecosystem is undefined, don't error out.a

Also, switch to using typescript with redux for better type checking.

### CODE REVIEW

File: .github/workflows/test-typescript.yml [hunk 1]
```diff
@@ -24,7 +24,7 @@ jobs:
       fail-fast: false
       matrix:
         os: [ ubuntu-latest, macos-latest, windows-latest]
-        node-version: [ 16, 18 ]
+        node-version: [ 16 ]
```
Comment (janpieterz) [inline .github/workflows/test-typescript.yml:27]: We don't want to confirm 18 compatibility?

File: browser/react-wallet/package.json [hunk 2]
```diff
@@ -46,9 +47,12 @@
     ]
   },
   "devDependencies": {
+    "@redux-devtools/extension": "3.2.2",
     "autoprefixer": "9.8.8",
     "postcss": "7.0.39",
     "postcss-loader": "4.3.0",
-    "tailwindcss": "npm:@tailwindcss/postcss7-compat@^2.2.4"
+    "redux-devtools-extension": "^2.13.9",
```
Comment (janpieterz) [inline browser/react-wallet/package.json:54]: Your friend renovate will have to PR for these using a non-specific version (`^`)

File: browser/react-wallet/src/App.tsx [hunk 3]
```diff
@@ -7,39 +7,40 @@ import PrivateRoute from './components/PrivateRoute';
...
+import {AppProps, AppState} from "./types";
+import {Dispatch} from "redux";
 
-function App(props) {
+function App(props: AppProps) {
   return (
     <div className="min-h-screen bg-gray-100">
       <Navbar className="fixed w-full bg-white shadow-md">
-        <Navbar.Logo href="/#">
+        <Navbar.Logo href="/#" className={""}>
```
Comment (janpieterz) [inline browser/react-wallet/src/App.tsx:17]: Empty classname? Any reason?

File: browser/react-wallet/src/App.tsx [hunk 4]
```diff
@@ -7,39 +7,40 @@ import PrivateRoute from './components/PrivateRoute';
...
           <TrinsicLogo className="h-12" />
         </Navbar.Logo>
         <Navbar.List className="justify-between">
-            {props.loggedIn && <Navbar.Item onClick={props.logout}>Sign Out</Navbar.Item> }
-            {props.loggedIn && <Navbar.Item>Ecosystem: {props.ecosystem.name}</Navbar.Item> }
+            {props.loggedIn && <Navbar.Item className={""} onClick={props.logout}>Sign Out</Navbar.Item> }
```
Comment (janpieterz) [inline browser/react-wallet/src/App.tsx:21]: Empty classname? Any reason?

File: browser/react-wallet/src/App.tsx [hunk 5]
```diff
@@ -7,39 +7,40 @@ import PrivateRoute from './components/PrivateRoute';
...
+            {props.loggedIn && <Navbar.Item className={""}>Ecosystem: {props.ecosystem?.name}</Navbar.Item> }
```
Comment (janpieterz) [inline browser/react-wallet/src/App.tsx:22]: Empty classname? Any reason?

**Review by janpieterz — COMMENTED** [review]
Very tough to review given the migration to TS in a lot of places.

File: .github/workflows/test-typescript.yml (hunk 1 above)
Comment (fundthmcalculus) [reply .github/workflows/test-typescript.yml:27]: 18 doesn't work with react wallet right now, I haven't had time to fix it yet.

File: browser/react-wallet/src/App.tsx (hunk 3 above)
Comment (fundthmcalculus) [reply browser/react-wallet/src/App.tsx:17]: Typescript warning, made explicit.

File: browser/react-wallet/src/App.tsx (hunk 4 above)
Comment (fundthmcalculus) [reply browser/react-wallet/src/App.tsx:21]: Typescript warning, made explicit.

File: browser/react-wallet/src/App.tsx (hunk 5 above)
Comment (fundthmcalculus) [reply browser/react-wallet/src/App.tsx:22]: Typescript warning, made explicit.

File: browser/react-wallet/src/actions/index.ts [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+            LoginRequest.fromPartial({
+                email,
+            })
+        );
+
+        dispatch({
+            type: LOGIN,
+            user: {name, email},
+            profile: undefined,
+            loginResponse: loginResponse
```
Comment (fundthmcalculus) [inline browser/react-wallet/src/actions/index.ts:47]: send the loginresponse / challenge correctly

File: browser/react-wallet/src/components/Button.js [hunk 7]
```diff
@@ -1,40 +1,54 @@
...
+
+    return (
+        <button className={baseClass} {...props} disabled={isBusy || disabled}>
+          <>
+            {children}
+            {isBusy && <Spinner className="inline-flex ml-2 w-4 h-4"/>}
+          </>
+        </button>
+    );
   }
```
Comment (fundthmcalculus) [inline browser/react-wallet/src/components/Button.js:39]: Make explicit React component to remove typescript warnings

File: browser/react-wallet/src/components/PrivateRoute.js [hunk 8]
```diff
@@ -2,7 +2,7 @@ import React from 'react';
 import { connect } from 'react-redux';
 import { Navigate } from "react-router-dom";
 
-function PrivateRoute({ children, loggedIn }) {
+function PrivateRoute({ children, loggedIn = true }) {
```
Comment (fundthmcalculus) [inline browser/react-wallet/src/components/PrivateRoute.js:5]: Default route is we are logged in

File: browser/react-wallet/src/index.tsx [hunk 9]
```diff
@@ -14,17 +14,17 @@ import App from './App';
...
 
 const persistedReducer = persistReducer(persistConfig, rootReducer);
-const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
-const middlewareEnhancer = applyMiddleware(thunkMiddleware);
-const enhancer = composeEnhancers(middlewareEnhancer);
-const store = createStore(persistedReducer, initialState, enhancer);
+const store = createStore(persistedReducer, initialState, composeWithDevTools(
+    applyMiddleware(thunkMiddleware)
+));
 let persistor = persistStore(store);
```
Comment (fundthmcalculus) [inline browser/react-wallet/src/index.tsx:28]: This fixes dev tools issues with typescript, modernized

File: browser/react-wallet/src/pages/CreateCredentialTemplatePage.tsx [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+    return {
+        schema: state.templates.schema
+    }
+}
+
+function mapDispatchToProps(dispatch: ThunkDispatch<ActionState, undefined, any>) {
+    return {
+        createTemplate: (name: string, fields: { [key: string]: TemplateField; }) => dispatch(createCredentialTemplate(name, fields))
+    }
+}
```
Comment (fundthmcalculus) [inline browser/react-wallet/src/pages/CreateCredentialTemplatePage.tsx:101]: We use thunking dispatch and state

File: browser/react-wallet/src/types.ts [hunk 11]
```diff
@@ -0,0 +1,19 @@
+export type AppProps = { loggedIn: boolean; logout: Function; ecosystem?: { name: string | undefined; }; }
+export type AppState = { authentication: { loggedIn: boolean; }; ecosystems: { currentEcosystem: any; }; }
+
+export type ActionState = {
+    wallet: any; (): any; new(): any; authentication: {
+        user: { name: string, email: string }; challenge: Uint8Array; profile: string;
+    };
+}
```
Comment (fundthmcalculus) [inline browser/react-wallet/src/types.ts:8]: @MichaelEdwardBlack we should probably find a way to consolidate these two?

**Review by fundthmcalculus — COMMENTED** [review]
Some examples of the key changes.

File: browser/react-wallet/src/components/Navbar.js [hunk 12]
```diff
@@ -167,4 +167,10 @@ export class Navbar extends React.Component {
   }
 }
 
+Navbar.Navbar = undefined;
+
```
[the remark this answers is not in the document]
Comment (fundthmcalculus) [reply browser/react-wallet/src/components/Navbar.js:171]: Agreed.

**Review by MichaelEdwardBlack — APPROVED** [review]

**Pull request merged**